File: Linked List/206. Reverse Linked List.py
"""
206. Reverse Linked List
Linked List

Description:
Reverse a singly linked list.

Example:
Input: 1->2->3->4->5->NULL
Output: 5->4->3->2->1->NULL

Follow up:
A linked list can be reversed either iteratively or recursively. Could you implement both?

Similar Question:
Reverse Linked List II - Medium
Binary Tree Upside Down - Medium
Palindrome Linked List - Easy
"""

# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        
        """
        Method 1: Iterative method
                  Note, if we let temp = current and current = temp.next, then when we modify current, we also modify temp.
                  But, if we let temp = current.next, then the changes on current won't affect temp.
        """
        head2 = None
        current = head
        while current:
            temp = current.next
            current.next = head2
            head2 = current
            current = temp
        return head2
    
        
        """
        Method 2: Recursive method
                  Same idea with method 1.
        """
        self.head2 = None
        return self.CreateRev(head)
    
    def CreateRev(self, head):
        if head:
            temp = head.next
            head.next = self.head2
            self.head2 = head
            head = temp
        else:
            return
        self.CreateRev(head)
        return self.head2

    
        

File: Linked List/206. Reverse Linked List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

// Method 1: Iterative
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode *newHead = nullptr;
        ListNode *cur;
        ListNode *tmp;
        cur = head;
        while(cur!=NULL){
            tmp = cur->next;
            cur->next = newHead;
            newHead = cur;
            cur = tmp;
        }
        return newHead;
    }
};

// Method 2: Recursive
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if (!head || !head -> next){
            return head;
        } 
        ListNode* ans=reverseList(head->next);
        head->next->next=head;
        head->next=NULL;
        return ans;
    }
};
